### ISSUE TITLE
[Clang][concepts] bug or feature? requires clause differs because of a pair of bracket

### ISSUE THREAD

**Issue by FeignClaims** [llvm/llvm-project#60377]
Like the example below, in order to help clang-format format requires clause correctly, I used to add a pair of bracket. Now I changed my style. When I tried to change my code, the unexpected error happened.

```cpp
#include <concepts>

template <typename T>
  requires std::integral<T>
struct [[nodiscard]] Widget;

template <typename T>
  requires(std::integral<T>)  // error: Requires clause differs in template redeclaration
struct [[nodiscard]] Widget {
  T value;
};

[[nodiscard]] auto main() -> int {
  Widget<int> widget{1};
  return widget.value;
}
```

I tested this with different compilers. It turns out:

- [error - x86_64 clang (trunk)](https://godbolt.org/z/6xTGY77cz)
- [pass - x86_64 gcc (trunk)](https://godbolt.org/z/xoM1M8Mvr)
- [error - x64 msvc v19.latest](https://godbolt.org/z/dcrYo3q7f)

Is this a bug or expected behaviour in standard?

**Comment by shafik** [comment]
I believe clang/MSVS are correct here, this is ill-formed no diagnostic required, see [temp.const.atomic p2](https://timsong-cpp.github.io/cppwp/temp.constr.atomic#2):

> ... As specified in [[temp.over.link]](https://timsong-cpp.github.io/cppwp/temp.over.link), if the validity or meaning of the program depends on whether two constructs are equivalent, and they are functionally equivalent but not equivalent, the program is ill-formed, no diagnostic required[.](https://timsong-cpp.github.io/cppwp/temp.constr.atomic#2.sentence-3) ...

**Comment by FeignClaims** [comment]
@shafik I see. Thanks.

**Comment by fahadnayyar** [comment]
@shafik @FeignClaims  What about the first bullet in the Constraint normalization section at this cpp referencee page: https://en.cppreference.com/w/cpp/language/constraints ?

`The normal form of an expression (E) is the normal form of E;`

**Comment by shafik** [comment]
@erichkeane if I am reading it correctly, it requires them constraints to be textually equivalent if they are functionally equivalent. Is this the correct reading?

**Comment by erichkeane** [comment]
That is my understanding, yes.

**Comment by shafik** [comment]
CC @hubert-reinterpretcast as C++ conformance code owner, what do you say?

**Comment by FeignClaims** [comment]
Apologize for the too early close. I noticed [temp.constr.atomic p1](https://timsong-cpp.github.io/cppwp/temp.constr#atomic-1) says:

> [Note [1](https://timsong-cpp.github.io/cppwp/temp.constr#atomic-note-1): Atomic constraints are formed by [constraint normalization](https://timsong-cpp.github.io/cppwp/temp.constr#def:constraint,normalization)[.](https://timsong-cpp.github.io/cppwp/temp.constr#atomic-1.sentence-2) E is never a [logical and expression](https://timsong-cpp.github.io/cppwp/expr.log.and) nor a [logical or expression](https://timsong-cpp.github.io/cppwp/expr.log.or)[.](https://timsong-cpp.github.io/cppwp/temp.constr#atomic-1.sentence-3) — end note]

So according to the quotation by @fahadnayyar (aka [temp.constr.normal p1](https://timsong-cpp.github.io/cppwp/temp.constr#normal-1)), I now agree the bracket should be fine.

**Comment by fahadnayyar** [comment]
Test1.cpp:
```
template<typename>
concept B = true;

template<typename T>
requires (B<T>)
struct A;

template<typename T>
requires B<T>
struct A {};
```

Test2.cpp:
```
template<typename>
concept B = true;

template<typename T>
requires ((B<T>) && (B<T>))
struct A;

template<typename T>
requires B<T> && B<T>
struct A {};
```

Test3.cpp:
```
template<typename>
concept B = false;

template<typename>
concept C = true;

template<typename T>
requires B<T> && (B<T> || C<T>)
struct A;

template<typename T>
requires B<T> && B<T> || C<T>
struct A {};
```

Clang should accept Test1.cpp and Test2.cpp but not Test3.cpp. Is that understanding correct?
Also should we replace the textual equivalence checking with NormalizedConstraint equivalence checking here: https://github.com/llvm/llvm-project/blob/main/clang/lib/Sema/SemaConcept.cpp#L794?

@shafik @erichkeane @EugeneZelenko @FeignClaims @hubert-reinterpretcast

**Comment by erichkeane** [comment]
>>Clang should accept Test1.cpp and Test2.cpp but not Test3.cpp. Is that understanding correct?
I'm not sure! I'm hoping someone more familiar with the standards would be able to chime in (like hubert).  As far as where to make that change, i think that would be the right place.

**Comment by FeignClaims** [comment]
Sorry, by rereading the standard, I feel even more confused about this issue. I would prefer to hear [more people's opinions](https://stackoverflow.com/questions/75317938/for-concept-a-should-requires-a-and-requires-a-be-identical).

**Comment by hubert-reinterpretcast** [comment]
> Clang should accept Test1.cpp and Test2.cpp but not Test3.cpp. Is that understanding correct?

No. `Test1.cpp` and `Test2.cpp` are IFNDR.

Declaration matching is not based on associated constraints (it used to be, but then that was to support allowing redeclarations to vary in their use of abbreviated function template syntax back in the history of the Concepts TS).

**Comment by FeignClaims** [comment]
> > Clang should accept Test1.cpp and Test2.cpp but not Test3.cpp. Is that understanding correct?
> 
> No. `Test1.cpp` and `Test2.cpp` are IFNDR.
> 
> Declaration matching is not based on associated constraints (it used to be, but then that was to support allowing redeclarations to vary in their use of abbreviated function template syntax back in the history of the Concepts TS).

If I understand correctly, there's no change that should take place, and the only thing left is to close the issue, right?

**Comment by shafik** [comment]
Closing since clang is behaving as expected, actually all the compilers are technically correct since ill-formed no diagnostic required means that while the compiler is allowed to diagnose this, it is a best effort and not diagnosing this is also conforming behavior.
